### PR TITLE
Support Roact Classes & .tsx

### PR DESCRIPTION
**Pull request by Vorlias**
Work In Progress. 

Definitely do not use for any production stuff yet. Things are hugely subject to change.

Issue: https://github.com/roblox-ts/roblox-ts/issues/20
Definitions: https://github.com/Vorlias/rbx-roact/blob/master/index.d.ts (WIP as well)

### CODE REVIEW

**Comment by Vorlias** [comment]
Tags are going to change to full lowercase without `rbx` prefix. Might have to also shorten some of the GUI object names to make them less... ugly.

I might also remove `ui` from the front of `UILayout` objects, since it just adds to the names.

File: tests/package.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
 	"author": "",
 	"license": "MIT",
 	"dependencies": {
-		"rbx-types": "^1.0.0"
+		"rbx-roact": "~0.0.7",
```
Comment (osyrisrblx) [inline tests/package.json]:
```suggestion
		"rbx-roact": "^0.0.7",
```

File: src/class/Transpiler.ts [hunk 2]
```diff
@@ -117,6 +117,47 @@ const LUA_RESERVED_METAMETHODS = [
...
 
+const ROACT_ELEMENT_TYPE = "Roact.Element";
+const ROACT_COMPONENT_TYPE = "Roact.Component";
+const ROACT_PURE_COMPONENT_TYPE = "Roact.PureComponent";
+
+/**
+ * A list of lowercase names that map to Roblox elements for JSX
+ */
+const INTRINSIC_MAPPINGS: { [name: string]: string } = {
+	// aspectratio: 	"UIAspectRatioConstraint",
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

File: src/class/Transpiler.ts [hunk 3]
```diff
@@ -1698,6 +1924,17 @@ export class Transpiler {
 			return this.transpileConditionalExpression(node);
 		} else if (ts.TypeGuards.isTypeOfExpression(node)) {
 			return this.transpileTypeOfExpression(node);
+
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

File: src/class/Transpiler.ts [hunk 4]
```diff
@@ -1698,6 +1924,17 @@ export class Transpiler {
...
+		} else if (ts.TypeGuards.isJsxExpression(node)) {
+
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

File: src/class/Transpiler.ts [hunk 5]
```diff
@@ -1698,6 +1924,17 @@ export class Transpiler {
...
+			return this.transpileExpression(node.getExpressionOrThrow());
+
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

File: src/class/Transpiler.ts [hunk 6]
```diff
@@ -1698,6 +1924,17 @@ export class Transpiler {
...
+		} else if (ts.TypeGuards.isJsxSelfClosingElement(node)) {
+			return this.transpileJsxSelfClosingElement(node);
+
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

File: src/class/Transpiler.ts [hunk 7]
```diff
@@ -1698,6 +1924,17 @@ export class Transpiler {
...
+		} else if (ts.TypeGuards.isJsxElement(node)) {
+			return this.transpileJsxElement(node);
+
```
Comment (osyrisrblx) [inline src/class/Transpiler.ts]:
```suggestion
```

Comment (Vorlias) [reply src/class/Transpiler.ts]: Removing that causes errors.

**Pull request merged**